src/concurrency/local_epoch.cpp

```diff
@@ -0,0 +1,118 @@
+//===----------------------------------------------------------------------===//
+//
+//                         Peloton
+//
+// centralized_epoch_manager.cpp
+//
+// Identification: src/concurrency/local_epoch.cpp
+//
+// Copyright (c) 2015-16, Carnegie Mellon University Database Group
+//
+//===----------------------------------------------------------------------===//
+
+#include "concurrency/decentralized_epoch_manager.h"
+
+
+namespace peloton {
+namespace concurrency {
+
+  bool LocalEpoch::EnterEpoch(const uint64_t epoch_id) {
+    epoch_lock_.Lock();
+    // if this thread is never used or GC'd
+    if (epoch_lower_bound_ == UINT64_MAX) {
+      epoch_lower_bound_ = epoch_id - 1;
+    } else if (epoch_lower_bound_ >= epoch_id) {
+    // epoch_lower_bound_ has already been updated by the GC.
+    // have to grab a newer epoch_id.
+      epoch_lock_.Unlock();
+      return false;
+    } 
+    // check whether the corresponding epoch exists.
+    if (epoch_map_.find(epoch_id) == epoch_map_.end()) {
+
+      std::shared_ptr<Epoch> epoch_ptr(new Epoch(epoch_id, 1));
+
+      epoch_queue_.push(epoch_ptr);
+      epoch_map_[epoch_id] = epoch_ptr;
+
+    } else {
+      epoch_map_.at(epoch_id)->txn_count_++;
+    }
+
+    epoch_lock_.Unlock();
+    return true;
+  }
+  
+  // the input parameter is the global read-only epoch id.
+  void LocalEpoch::EnterEpochRO(const uint64_t epoch_id) {
+    epoch_lock_.Lock();
+    // if this thread is never used or GC'd
+    if (epoch_lower_bound_ == UINT64_MAX) {
```

Comment:
Why do we also need to update epoch_lower_bound_ here when creating RO txn? I assume that the epoch_id passed to this function is of a "stale" epoch.